### PR TITLE
nxpmicro-mfgtools: init at 1.3.191

### PR DESCRIPTION
**Pull request by bmilanov**
Add nxpmicro-mfgtools to nixpkgs

From the project's homepage [1]:

    Freescale/NXP I.MX Chip image deploy tools.

The project's only binary and library at the moment is UUU. NXP
might add more in the future, so making the package name:

    nxpmicro-mfgtools

instead of:

    nxp-uuu

or something similar.

Add a preConfigure step based on bbigras' suggestion.

[1]: https://github.com/NXPmicro/mfgtools

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by bbigras — CHANGES_REQUESTED** [review]
I think it would be cleaner that way.

EDIT: derp, I didn't add the ~requested~ suggested changes.

File: pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix
Comment (bbigras) [inline pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix:12]:
```suggestion
  pname = "nxpmicro-mfgtools";
  version = "1.3.191";
```

File: pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix
Comment (bbigras) [inline pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix:13]:
```suggestion
  pname = "nxpmicro-mfgtools";
  version = "1.3.191";
```

File: pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix
Comment (bbigras) [inline pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix:29]:
```suggestion
    rev    = "uuu_${version}";
```

File: pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix
Comment (bbigras) [reply pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix:13]: According to https://nixos.org/nixpkgs/manual/#sec-using-stdenv , pname with set the name with the right version.

**Comment by bbigras** [comment]
You didn't check "Tested using sandboxing". Are you sure you are not using sandboxing? It's enabled by default in NixOS.

You can check by running: `nix-shell -p nix-info --run "nix-info -m"`

File: pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix
Comment (bbigras) [inline pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix:23]: Are you sure it's Linux only? The project's readme says:
> The real cross platform. Linux, Windows, MacOS(not test yet)

Maybe try `platforms.all`.

File: pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix
Comment (bbigras) [inline pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix:36]:
```suggestion
  nativeBuildInputs = [ cmake pkg-config ];
  buildInputs       = [ bzip2 libusb1 libzip openssl ];
```

> build time only dependencies are declared in `nativeBuildInputs`

https://nixos.org/nixpkgs/manual/#reviewing-contributions-new-packages

I'm pretty sure `pkg-config` is only needed at build-time.

File: pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix
Comment (bbigras) [inline pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix:16]: > For brevity, don’t repeat the name of package — just describe what it does

Check https://nixos.org/nixpkgs/manual/#sec-standard-meta-attributes

I think a description of what the program does would be better.

File: pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix
Comment (bmilanov) [reply pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix:23]: It says it's cross-platform, but I wanted to add only the platform that I am testing on.

Should I add all stated working, or should I add only tested platforms?

Comment (bbigras) [reply pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix:23]: I don't know what the NixOS policy exactly is on this but maybe it's fine if we put "all" and the ofborg bot doesn't complain while building for all the platforms it supports.

File: pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix
Comment (bmilanov) [reply pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix:16]: I wonder if it's better to get the contents of the long description in this field. Then think about what to add in the long description itself.

**Comment by bmilanov** [comment]
Apologies for the multiple commits, I just now saw that I can batch suggestions into one.

File: pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix
Comment (bmilanov) [reply pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix]: How about this: 5f5d2ee ?

File: pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix
Comment (bmilanov) [reply pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix]: 13f3760 fingers crossed

**Comment by bmilanov** [comment]
> You didn't check "Tested using sandboxing". Are you sure you are not using sandboxing? It's enabled by default in NixOS.
> 
> You can check by running: `nix-shell -p nix-info --run "nix-info -m"`

Seems like I have:

    - sandbox: `yes`

**Comment by bbigras** [comment]
@GrahamcOfBorg build zenith

**Comment by bbigras** [comment]
Ops I pasted the wrong command.

**Comment by bbigras** [comment]
@GrahamcOfBorg build nxpmicro-mfgtools

**Comment by bbigras** [comment]
Result of `nixpkgs-review pr 91634` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
<br>- nxpmicro-mfgtools
</details>

**Review by bbigras — APPROVED** [review]

**Comment by bbigras** [comment]
I don't know if it matters for new packages but here's the closure size:

❯ nix path-info -S -h /nix/store/fkfi4ainw1iq49zp2jgpmmd0dyqrhixd-nxpmicro-mfgtools-1.3.191
/nix/store/fkfi4ainw1iq49zp2jgpmmd0dyqrhixd-nxpmicro-mfgtools-1.3.191	  47.5M

File: pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix
Comment (danieldk) [inline pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix:32]: Canonical order: `meta` comes last in `mkDerivation`.

File: pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix
Comment (danieldk) [inline pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix:28]: Unquoted URLs are deprecated:

```suggestion
    homepage = "https://github.com/NXPmicro/mfgtools";
```

File: pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix
Comment (danieldk) [inline pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix:16]:
```suggestion
    description = "Freescale/NXP I.MX chip image deploy tools";
```

File: pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix
Comment (danieldk) [inline pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix:44]: `nativeBuildInputs` and `buildInputs` go between `src` and `preConfigure`.

File: pkgs/top-level/all-packages.nix
Comment (danieldk) [inline pkgs/top-level/all-packages.nix:5593]:
```suggestion
  nxpmicro-mfgtools = callPackage ../applications/science/electronics/nxpmicro-mfgtools { };
```

Many other derive flashing utilities (e.g. `esptool`, `stm32flash`) are in `development/tools/misc`, so that may be the more logical location.

File: pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix
Comment (danieldk) [inline pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix:12]: Maybe this should be just be called `mfgtools`? Not 100% sure though ;).

**Review by danieldk — CHANGES_REQUESTED** [review]
Thank you for contributing to `nixpkgs`. I have added some more comments.

Could you squash the commits together? The first commit should add you to `maintainers.nix`, the second should add the derivation.

File: pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix
Comment (bmilanov) [reply pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix]: Hi, thanks for all the feedback!

I am not sure which way's better.

Usually I would add the "nxpmicro-" as I don't know how many mfgtools can pop up in the future (doubt there will be any more), with the assumption that adding "namespaces" later would be harder. I am really tempted to make it only "mfgtools" as it's less text to write.

Your call. I can amend the commit based on what you decide.

Comment (danieldk) [reply pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix]: Thanks for the changes. There was a small syntax error that caused evaluation to fail, so I fixed that. I also used `nixpkgs-fmt` to format the derivation.

I think the prefix may be better here. As you say, `mfgtools` is very generic. I don't think there will be another package with the same name, but with the prefix it's easier to recognize and quickly find.

Comment (danieldk) [reply pkgs/applications/science/electronics/nxpmicro-mfgtools/default.nix]: Whoops, seems like we are force-pushing at the same time. If you only fixed the syntax error, I can push my changes again.

**Comment by bmilanov** [comment]
Just after I pushed I saw that you've changed the comma style because of the build error.

I will back off so you can force push to the proper commit.

**Review by danieldk — APPROVED** [review]
Thanks, looks great now!

Result of `nixpkgs-review pr 91634` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
<br>- nxpmicro-mfgtools
</details>

Tested binary to work.

**Comment by bbigras** [comment]
Thanks @bmilanov for your contribution and perseverance.

Thanks @danieldk too.

**Pull request merged**
